Allow the userdata template to be replaced

snstanton opened the pull request:
# PR o'clock

## Description

This addresses issue #564 

Adds a `custom_userdata` property to `worker_groups` that can be used to override the full userdata template file.

### Checklist

- [x] `terraform fmt` and `terraform validate` both work from the root and `examples/*` directories
- [x] CI tests are passing
- [x] I've added my change to CHANGELOG.md and highlighted any breaking changes
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

snstanton commented:
I've attempted to run validate, but due to https://github.com/terraform-providers/terraform-provider-aws/issues/9989, the validate fails with:

> "region" is required but not set

Pretty sure this has nothing to do with this PR.

max-rocket-internet commented on data.tf:
```diff
@@ -89,8 +89,12 @@ EOF
+  count = local.worker_group_count
+  template = lookup(
+    var.worker_groups[count.index],
+    "custom_userdata",
```
I love it!

But can we change this key name? Since here we are not supplying the actual user data, we are supplying a template attribute for the `template_file` resource. So how about `userdata_template_file`?

snstanton replied:
Sure.  I'll make that change.

barryib commented:
If we want to add, shouldn't we add an extra variable (like `userdata_template_extra_args`) to pass to the template ? Because with this PR, you can change the template, but you can only use these variables https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/data.tf#L91-L114. What if your template need another variables ?

@snstanton @max-rocket-internet What do you think ?

snstanton commented:
That's a good question.  I actually ran into this in my use case.  What I wound up doing was expanding the template twice.  Once with my variables and then again after I pass it to the module.  I had to escape the references to the standard template vars so they would pass through the first expansion.  Your idea of passing an extra args map is interesting and would probably simplify this use case.

barryib commented on data.tf:
```diff
@@ -112,7 +116,8 @@ data "template_file" "userdata" {
     )
-  }
+    },
+  lookup(var.worker_groups[count.index], "userdata_template_extra_args", {}))
```
The default should be `local.workers_group_defaults["userdata_template_extra_args"]` and not `{}` => 

```hcl
lookup(
  var.worker_groups[count.index],
  "userdata_template_extra_args",
  local.workers_group_defaults["userdata_template_extra_args"]
)
```

barryib commented:
@snstanton thanks you for all your work. I just notice that  https://github.com/terraform-aws-modules/terraform-aws-eks/pull/555 is now merded ans it adds windows support.

Can you please add the same logic in the windows userdata template ? Thanks.

snstanton commented:
I wonder if it would it be simpler to just make `platform` available inside the template context so a single template can contain platform specific sections if needed?  It's a bit awkward to have to supply two template files but only one extra arguments map.

snstanton commented:
I've added `platform` to the template expansion context.  If you still want a separate `userdata_windows_template_file`, let me know and I'll add that as well.

I have also added this functionality to the `launch_template_userdata` template since it was largely the same.

barryib commented:
I see that @max-rocket-internet requested some changes. Waiting for his feedback before merging.
